### DISCUSSION TITLE
API search query syntax

### DISCUSSION THREAD

**Discussion by robintherocker** [lnx-search/lnx#157] [Q&A] [answered]
Hi!

I'm having a hard time finding any _useful_ documentation about the query syntax.

Would be very nice to get some documentation on how to query an index beyond the super basic query, and how to use operands (AND, OR). Somewhere else, I found this example:
```
{
  "query": [
     { "occur": "must", "normal": "my tantivy parser query here" },   <-- Filter
     { "occur": "must", "fuzzy": "My nice fuzzy search here" },   <--- User search
  ]
}
```

When I test the above example I get all results that matches line a OR line b, so the first line is not actually a filter as stated in the comment.
After hours of messing around with the API, all I found was that I can make AND or OR statements like this:
```
{
  "query": [
    {
        "normal": {"ctx": "filecontent:somekeyword AND fileextension:csv"}
    }
     
 ],
  "limit": 10,
  "offset": 0,
  "order_by": "score",
  "sort": "desc"
}
```

It seems to look at lot like SQL syntax, so I can also add multiple operands like this:
```
{
  "query": [
    {
        "normal": {"ctx": "fileextension:xls AND (content:KeywordA OR content:KeywordB) AND customer:somecustomername"}
    }
     
 ],
  "limit": 10,
  "offset": 0,
  "order_by": "score",
  "sort": "desc"
}
```

I would love if someone could show more advanced usage, using wildcards and NOT statements... Anyone? @ChillFish8 ? :D 
Or explain to me if I misunderstood how to build queries...

As an example, Excel documents could be .xls or .xlsx and even other file extensions... Without being able to use a wildcard, I'll need to search with a query like this:

```
{
  "query": [
    {
        "normal": {"ctx": "(fileextension:xls OR fileextension:xlsx OR fileextension:xlsm) AND (content:KeywordA OR content:KeywordB) AND customer:somecustomername"}
    }
     
 ],
  "limit": 10,
  "offset": 0,
  "order_by": "score",
  "sort": "desc"
}
```

Adding a star (fileextension:xls*) returns no results, while fileextension:* returns ALL file type results. Seems like either I'm missing something, or the wildcard doesn't really work. 

My target users are VERY pleased with LNX-search, saying it's super fast! 🥇

**Comment by ChillFish8** [comment, marked as answer]
The `normal` query mode uses the Tantivy parser syntax: https://docs.rs/tantivy/0.18.1/tantivy/query/struct.QueryParser.html

Unfortunately, ATM 0.9 uses an old version of tantivy which means you can't do your example like `fileextension:xls*` (which is would be a prefix match in newer versions of tantivy)
